### PR TITLE
Task 3

### PR DESCRIPTION
**Pull request by MateuszKantyka**

### CODE REVIEW

File: app/views/users/index.html.erb [hunk 1]
```diff
@@ -6,7 +6,7 @@
     <td><%= sort_link(@q, :email) %></td>
     <td><%= sort_link(@q, :is_male, 'Gender') %></td>
     <td><%= sort_link(@q, :birthday) %></td>
-    <td></td>
+    <td><%= sort_link(@q, :interests) %></td>
```
Comment (wojtodzio) [inline app/views/users/index.html.erb]: This sorting does not work 😿 But I think that we can skip sorting for that column, it's not required 😄

File: app/views/users/index.html.erb [hunk 2]
```diff
@@ -15,6 +15,8 @@
         <td><%= user.email %></td>
         <td><%= user.is_male ? 'male' : 'female' %></td>
         <td><%= user.birthday %></td>
+        <td><%= user.interests.map { |interest| interest.name }.join(', ')  %></td>
```
Comment (wojtodzio) [inline app/views/users/index.html.erb]: `user.interests.map` results in the following query:
```SQL
SELECT "interests".* FROM "interests" WHERE "interests"."user_id" = ?
```
In other words, it:
- fetches all interests data (interests.*),
- for each of them, it instantiates a separate active record object,
- then it takes only `name` from it
-  and then it disregards those AR objects as not needed anymore.

A much more performant option would be to use the following query:
```SQL
SELECT "interests"."name" FROM "interests" WHERE "interests"."user_id" = ?
```
So, to fetch only needed data (names), and use them without instantiating unnecessary AR objects.
[You can use `pluck` for that](https://apidock.com/rails/ActiveRecord/Calculations/pluck).

File: db/seeds.rb [hunk 3]
```diff
@@ -6,10 +6,16 @@
...
                password: Faker::Internet.password(8),
                is_male: Faker::Boolean.boolean,
                admin: false,
                birthday: Faker::Date.birthday(18, 65))
 end
+
+7.times do
+  Interest.create!(name: Faker::ProgrammingLanguage.name,
+                   type: 'Work',
+                   user_id: Faker::Number.between(1,3))
```
Comment (wojtodzio) [inline db/seeds.rb]: It's a bad practice to assume what id a record will have.
Think of that in that way: what if
- someone will populate his DB with seeds,
- then he will delete all users from his DB,
- and then he will try to use seeds again?

Ids of the newly created users (those from running seeds the second time) will be 3 and 4, and as so creating interests will fail, as it will try to assign numbers between 1 and 3 for theirs user_ids.
It would be much better to use ids of the real users from the DB here.
And performance in seeds is not important at all, so you can use the simplest approach available (e.g. `User.all.shuffle.first`).

File: spec/factories/interests.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+FactoryBot.define do
+  factory :interest do
+    name 'Ruby'
+    type 'Work'
```
Comment (wojtodzio) [inline spec/factories/interests.rb:4]: It's a good idea to have all factories working without specifying any parameters - in that way, you can always use them by overriding only those parameters, that are important to the specific test.
But here it will fail because of the missing `user_id`.

File: spec/features/admin_spec.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
+require 'rails_helper'
+
+RSpec.feature 'Admin' do
+  scenario 'admin can delete user' do
+    admin = create(:user, admin: true)
+    create(:user, email: 'example-user@mail.com')
+    sign_in admin
+
+    visit root_path
+    find(:xpath, "//a[@href='/users/2']").click
```
Comment (wojtodzio) [inline spec/features/admin_spec.rb]: There is a high probability, that this test will fail in the future for some reasons:
- Adding any other link to '/users/2' will make it fail, as there will more than one match - and we use that path for show, so it will probably be there sooner or later
- Users path can change - always use `user_path(user)` instead
- You are assuming the id of the user (2), and small change to this test or to the way records are cleaned between tests can break it

I've just explained it further in the mail.

File: spec/features/admin_spec.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+  scenario 'admin can delete user with interests' do
+    admin = create(:user, admin: true)
+    user = create(:user, email: 'example-user@mail.com')
+    create(:interest, user_id: user.id)
+    create(:interest, user_id: user.id)
+    sign_in admin
+
+    visit root_path
+    find(:xpath, "//a[@href='/users/2']").click
```
Comment (wojtodzio) [inline spec/features/admin_spec.rb]: Same here 😄

File: spec/models/interest_spec.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
+require 'rails_helper'
+
+RSpec.describe Interest do
+  it 'has none to begin with' do
+    expect(Interest.count).to eq 0
```
Comment (wojtodzio) [inline spec/models/interest_spec.rb]: What is the reason for that test?

File: spec/models/interest_spec.rb [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+  end
+
+  it 'can by created with valid arguments' do
+    user = create(:user)
+    Interest.new(name: 'jokes', type: 'Hobby', user_id: user.id).save
+
+    expect(Interest.count).to eq 1
```
Comment (wojtodzio) [inline spec/models/interest_spec.rb]: Assuming the number of records in the DB is not a very good practice neither. I believe you can make interest factory work without additional parameters and use a spec with the FactoryBot lint instead of that one.
Something along lint_spec.rb
```ruby
RSpec.describe 'FactoryBot linter' do
  it 'verifies that all FactoryBot factories are valid' do
    FactoryBot.lint
  end
end
```
It will try to create a record for each factory with the default parameters, and it will fail if any of the cannot be created - hence making the spec above obsolete 😄

File: spec/models/user_spec.rb [hunk 9]
```diff
@@ -0,0 +1,32 @@
+require 'rails_helper'
+
+RSpec.describe User do
+  it 'has none to begin with' do
+    expect(User.count).to eq 0
```
Comment (wojtodzio) [inline spec/models/user_spec.rb]: Same as above, what does that test tests, and why is it needed?

File: spec/models/user_spec.rb [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  end
+
+  it 'has a valid Factory' do
+    create(:user)
+
+    expect(User.count).to eq 1
```
Comment (wojtodzio) [inline spec/models/user_spec.rb]: Same as above, use FactoryBot linter instead

File: spec/models/user_spec.rb [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+  end
+
+  it 'can by created with valid arguments' do
+    User.new(email: 'example@mail.com',
+             password: 'qwerty',
+             is_male: true,
+             admin: false).save
+
+    expect(User.count).to eq 1
```
Comment (wojtodzio) [inline spec/models/user_spec.rb]: Same as above

File: app/views/users/index.html.erb [hunk 12]
```diff
@@ -6,7 +6,7 @@
     <td><%= sort_link(@q, :email) %></td>
     <td><%= sort_link(@q, :is_male, 'Gender') %></td>
     <td><%= sort_link(@q, :birthday) %></td>
```
Comment (wojtodzio) [inline app/views/users/index.html.erb]: In the specification, the listing shows users age, not birthday. You should probably calculate age based on the birthday. Try not to put that logic into view :)

**Review by wojtodzio — CHANGES_REQUESTED** [review]

File: spec/helpers/user_helper_spec.rb [hunk 13]
```diff
@@ -0,0 +1,26 @@
+require 'rails_helper'
+include ActiveSupport::Testing::TimeHelpers
+
+RSpec.describe UsersHelper do
+  describe '#user_age' do
+    context "when today's date is 2018-01-30 and user was born in 1994-08-21" do
```
Comment (wojtodzio) [inline spec/helpers/user_helper_spec.rb]: From your contexts/its I know, that the `user_age` helper:
- returns 23 if called 2018-01-30 with user born 1994-08-2,
- returns 24 if called 2018-01-30 with user born 1994-08-21,
- and I have no idea how it behaves when called on other dates and with different users

Of course, I can guess it as it's quite obvious now in that context, but ideally, specs should work as a documentation of your code. If you are not able to cover all possible cases with a separate contexts, then maybe that's not a way to go.
How about doing something as simple, as a single `it` with something along the lines "it returns the current age in years of a given user" - that describes exactly what that method does, and it can be quite easily tested.
If you want to make sure it works for more cases than you can simply create more than one user in the single spec and test it for all of them.
And if you want to test edge cases, you can wrap that single it in a context (e.g. "when user have a valid birthday set"), and add a separate one for the edge case (e.g. "when user have a birthday in the future").

File: app/helpers/users_helper.rb [hunk 14]
```diff
@@ -0,0 +1,6 @@
+module UsersHelper
+  def user_age(user)
+    now = Time.now.utc.to_date
```
Comment (wojtodzio) [inline app/helpers/users_helper.rb]: I think it's easier to use just `Date.current` instead

File: app/helpers/users_helper.rb [hunk 15]
```diff
@@ -0,0 +1,6 @@
...
+    now.year - user.birthday.year - ((now.month > user.birthday.month || (now.month == user.birthday.month && now.day >= user.birthday.day)) ? 0 : 1)
```
Comment (wojtodzio) [inline app/helpers/users_helper.rb]: That line has a lot of logic in it and it's really hard to follow it.
It would be nice to make it more readable - try splitting it to multiple lines, assigning parts of it into well-named variables etc., please 😄

**Review by wojtodzio — APPROVED** [review]
Just a few really small things left, almost there 😄

**Review by wojtodzio — CHANGES_REQUESTED** [review]
I missclicked and approved by accident, it was meant to be "request changes" 😆

**Review by wojtodzio — APPROVED** [review]
